text/0000-export-visibility.md

```diff
@@ -0,0 +1,729 @@
+/// This is a public (`pub`) Rust function - it may be called from other Rust
+/// crates.
+///
+/// This function may internally (say, as an implementation detail) call
+/// `fn internal_helper_called_from_rust_or_cpp` above.  If this public function
+/// gets inlined into another `dylib` then the call to the internal helper
+/// will cross `dylib` boundaries - this will **not** work if the internal
+/// helper is hidden from dynamic linking.
+#[inline]
+pub fn public_function() {
+    internal_helper_called_from_rust_or_cpp()
+}
+```
+
+### Potential answers
+
+The following options have been identified so far as a potential way for
+answering the `dylib`-vs-`hidden`-visibility problem:
+
+* Don't stabilize (or don't support at all) `#[export_visibility = "hidden"]`
+  but still support other visibilities
+* Support `#[export_visibility = "hidden"]`, but
+    - Document that `hidden` visibility may break linking of `dylib`s
+      (see the "Hidden visibility" section in the guide-level explanation above)
+    - Document a recommendation that reusable crates shouldn't use a hardcoded
+      visibility (see the "Choosing the right visibility" section in the
+      guide-level explanation above)
+* Investigate if cross-`dylib`-inlining can (should?) be avoided if the inlined
+  code ends up calling a hidden symbol from the other crate.
+
+## Cross-platform behavior
+[cross-platform-behavior]: #cross-platform-behavior
+
+We don't really know
+whether the `hidden` / `protected` / `interposable` visibilities
+make sense across different target platforms and/or map to distinct entities
+(see
+[a Zulip question here](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/.60.23.5Bexport_visibility.20.3D.20.2E.2E.2E.5D.60.20attribute.20compiler-team.23881/near/522491140)).
+
+One weak argument is that these visibilities are supported by LLVM and Clang, so hopefully
+they would also make sense for Rust:
+
+* **LLVM**: Those visibilities are ultimately mapped from
+[`rustc_target`'s `SymbolVisibility`](https://github.com/rust-lang/rust/blob/81a964c23ea4fe9ab52b4449bb166bf280035797/compiler/rustc_target/src/spec/mod.rs#L839-L843),
+through
+[`rustc_middle`'s `Visibility`](https://github.com/rust-lang/rust/blob/81a964c23ea4fe9ab52b4449bb166bf280035797/compiler/rustc_middle/src/mir/mono.rs#L396-L407),
+and into
+[`rustc_codegen_llvm`'s `Visibility`](https://github.com/rust-lang/rust/blob/81a964c23ea4fe9ab52b4449bb166bf280035797/compiler/rustc_codegen_llvm/src/llvm/ffi.rs#L153-L160).
+So all the values make some sense at
+[the LLVM level](https://llvm.org/docs/LangRef.html#visibility-styles).
+* **Clang** and **GCC** support those 3 visibilities
+  (see the "Parity with C++" subsection in the "Motivation" section above).
+
+OTOH, ideally we would somehow check what happens on some representative subset
+of target platforms (maybe: Posix, Windows, Wasm?):
+
+* TODO: what exactly do we want to verify on these target platforms?
```

Comment:
I don't think we can merge the RFC without any due diligence here. The RFC should demonstrate that the parameters it gives for each option can be implemented for at least our Tier 1 platforms, or specify where there are exceptions or specific unknowns.